### DISCUSSION TITLE
Rule with publish counter value

### DISCUSSION THREAD

**Discussion by pistoletov1974** [arendst/Tasmota#20942] [Q&A]
Hi ! I just want to publish counter value if button unpressed. This is  my rule:
Rule1  ON POWER1#State == OFF DO Publish garage/counter Counter1 ENDON.
But it publish "Counter1" string in topic.  What is correct syntax to publish counter value in topik

**Comment by lalo-uy** [comment]
You have to capture counter1 on a var when it moves.

On Counter1 do var1 = %value% endone
ON POWER1#State == OFF DO Publish garage/counter var1 endone

**Reply by pistoletov1974** [reply]
i received var1 string  in payload.
My new rule:
stat/tasmota_kiosk/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":102,"Free":409,"Rules":"On Counter1 do var1 = %value% ENDON ON POWER1#State == OFF DO Publish garage/counter var1 ENDON"}}

**Reply by sfromis** [reply]
That wont't work, as counter updates cause no publish of that JSON format to pick up, and there are syntax errors.

Here's a working way:
```
on Power1#state==0 do Backlog Counter1 endon
on Counter1 do publish garage/counter %value% endon
```
First rule will publish the counter value to the standard topics. 
2nd rule, which is not really needed, will pick that up, and do another publish.
The use of `Backlog` in the first rule is to have the result available to the next.

You could have a rule trigger like `Counter#c1` which will pick up the counter value each cycle, updated or not, which you then could mitigate by saving the value, and only triggering if the value was different, but the above is a more straightforward approach.

**Reply by pistoletov1974** [reply]
Thank you, @sfromis but this way does not work to:
> 16:29:21.024 MQT: stat/tasmota_kiosk/POWER1 = OFF
> 16:29:21.038 RUL: POWER1#STATE==0 performs "Backlog Counter1"
> 16:29:21.068 MQT: stat/tasmota_kiosk/RESULT = {"Counter1":5093}
> 16:29:21.085 RUL: COUNTER1 performs "publish garage/counter "
> 16:29:21.092 MQT: garage/counter =

 
As we can see empty value was published
This is rule:
`Rule1 ON Power1#state==0 DO  Backlog Counter1 ENDON ON Counter1 DO publish garage/counter %value% ENDON`

**Reply by sfromis** [reply]
Oops, I got my test mixed a bit up. Should've been:
`on Counter1#data do publish garage/counter %value% endon`
Needing that `#data` is a special quirk around command result JSON with single values.

**Reply by pistoletov1974** [reply]
@sfromis the problem is in different trigger (button) and published value (counter). If i use rule like this 
`on Counter1#data do publish garage/counter %value% endon`
 values published  many time when counter changed. Doe rules support  nested conditions?

**Reply by pistoletov1974** [reply]
@lalo-uy thank you after little correction yours rule work fine
Rule1  On Counter#c1 do var1 = %value% ENDON ON POWER1#State == OFF DO Publish garage/counter %var1% ENDON 
Little inconinience is that very often trigger counter#c1 fired. And console full of this messages

**Reply by sfromis** [reply]
Not clear what you mean by "different trigger"? When I change the counter value (by input gpio or command), I get exactly *one* triggering of the `Counter1#data` rule, nothing more. Of course, if you frequently switch on/off the `Power1#state`, you'll get a Counter1 command every time switching off, triggering the rule again, but that was just what you asked for. Without the log, I can't suggest what your issue is.

When it comes the to the `Counter#c1` strategy, I already described how you can deal with that trigger firing all the time, by saving the value in a var, and comparing to that in the trigger. Anyway, that method also has another side effect, the value you get can be a couple of seconds back in time.

**Reply by pistoletov1974** [reply]
@sfromis counter count pulses from some external device NOT from the switch.   When switch go to off state rule must send captured counter  to mqtt topic.  Trigger for sending mqtt is switch, but data to send is counter value.  "On  switch ==off  send  COUNTER value to mqtt topic" .  Counter input is external input from hall sensor.  Switc is a mechanical switch.-)  Counter and switch this is independent inputs.

**Reply by sfromis** [reply]
Do you mean Relay/Power when you say "switch"? In Tasmota, a "Switch" is an input pin (not related to counter inputs), while simple output pins typically are a "Relay". And a trigger on `Power1#state` will trigger on the power state change (no matter what caused it) flipping the Relay output. If you happen to have a Switch input pin resulting in the power state change, this is indeed not directly involved in what the counter is doing.

What Tasmota is doing is that transitions on the Counter input pin results in the counter value being incremented. It cannot trigger further actions by itself. With a rule triggering on `Counter#c1`, you get a delayed value in the sensor cycle, not every increment, but usually relatively close.

The point of having a `Power1#state` trigger to issue a `Counter1` command is that this will query the current counter value, resulting in a single result like `{"Counter1":106}` in JSON format, and then a `on Counter1` trigger will react to that JSON. It will not react to just the counter incrementing without messages appearing.

But still, as you did not post a log, I can't see what's actually happening.

**Reply by pistoletov1974** [reply]
@sfromis thank you for reply. Yes, correct: 

> If you happen to have a Switch input pin resulting in the power state change, this is indeed not directly involved in what the counter is doing.

Hmm...does it possible to publish mqtt with all sensors values immediately after POWER1#State == OFF in custom topic ? Now all my sensors data published only in teleperiod : 16:19:46.024 MQT: tele/tasmota_kiosk/SENSOR =.......

can i publish simple sensor value through console?  I try 
17:42:06.219 CMD: **publish kiosk/2 counter#c1**
17:42:06.225 MQT: kiosk/2 = counter#c1
But this does not work.

Now rule is:
`"On Counter#c1 do var1 = %value% ENDON ON POWER1#State == OFF DO Publish garage/counter %var1% ENDON"}}`


This is log:
16:18:06.615 RUL: COUNTER#C1 performs "var1 = 1494"
781 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:16:42.168 RUL: COUNTER#C1 performs "var1 = 1494"
16:16:42.176 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:16:44.448 MQT: cmnd/tasmota_kiosk/POWER1 = ON
16:16:44.501 MQT: stat/tasmota_kiosk/RESULT = {"POWER1":"ON"}
16:16:44.504 MQT: stat/tasmota_kiosk/POWER1 = ON
16:16:46.578 RUL: COUNTER#C1 performs "var1 = 1494"
16:16:46.587 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:16:51.103 RUL: COUNTER#C1 performs "var1 = 1494"
16:16:51.112 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:16:55.595 RUL: COUNTER#C1 performs "var1 = 1494"
16:16:55.603 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:16:59.967 RUL: COUNTER#C1 performs "var1 = 1494"
16:16:59.977 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:17:04.482 RUL: COUNTER#C1 performs "var1 = 1494"
16:17:04.490 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:17:08.895 RUL: COUNTER#C1 performs "var1 = 1494"
16:17:08.903 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:17:13.285 RUL: COUNTER#C1 performs "var1 = 1494"
16:17:13.294 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:17:17.680 RUL: COUNTER#C1 performs "var1 = 1494"
16:17:17.689 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
.....
....

16:20:23.757 MQT: stat/tasmota_kiosk/POWER1 = OFF
**16:20:23.773 RUL: POWER1#STATE == OFF performs "Publish garage/counter 1494.000"**
**16:20:23.778 MQT: garage/counter = 1494.000**
16:20:23.791 RUL: COUNTER#C1 performs "var1 = 1494"
16:20:23.798 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:20:28.206 RUL: COUNTER#C1 performs "var1 = 1494"
16:20:28.215 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}
16:20:32.623 RUL: COUNTER#C1 performs "var1 = 1494"
16:20:32.631 MQT: stat/tasmota_kiosk/RESULT = {"Var1":"1494.000"}

**Reply by pistoletov1974** [reply]
@lalo-uy , @sfromis 

> You have to capture counter1 on a var when it moves.
> On Counter1 do var1 = %value% endone

this rule works but too often it fires even when the value of counter does not change. What can I do to make the rule work only when the value changes?

```
14:18:08.229 CMD: rule1 On Counter#c1 do var1 = %value% ENDON
....
14:18:24.141 MQT: tele/tasmota_don1/SENSOR = {"Time":"2024-03-26T14:18:24","COUNTER":{"C1":530}}
14:18:27.145 RUL: COUNTER#C1 performs "var1 = 530"

every near 5 seconds rule fired

14:24:53.071 RUL: COUNTER#C1 performs "var1 = 530"
14:24:53.083 MQT: stat/tasmota_don1/RESULT = {"Var1":"530.000"}
14:24:57.470 RUL: COUNTER#C1 performs "var1 = 530"
14:24:57.482 MQT: stat/tasmota_don1/RESULT = {"Var1":"530.000"}
14:25:01.858 RUL: COUNTER#C1 performs "var1 = 530"
14:25:01.866 MQT: stat/tasmota_don1/RESULT = {"Var1":"530.000"}
14:25:06.242 RUL: COUNTER#C1 performs "var1 = 530"
14:25:06.253 MQT: stat/tasmota_don1/RESULT = {"Var1":"530.000"}}
```
